### PR TITLE
net:gptp: Always use GM PRIO root system id for announce messages

### LINKED ISSUE #66646: Best Master Clock identification in gPTP using multiple ports
The following problem occurred while running PTP on a multi-port target (RENESAS RZT2M) with two ports enabled but it shoul dalso happen with other Multi-Port devices. BMCA was configured in a way that the device (Dev-B) acts as bridge with a master port (Port2) and a slave port (Port1). Slave port connects to the GM (Dev-A). Master port connects to another device (Dev-C). The devices BMCA algorithms  have been configured using the according priorities:
 - Dev-A: 100
 - Dev-B: 200
 - Dev-C: 150

 That should result in the following PTP-Tree:
<pre>
 ┌─────────┐                ┌─────────┐              ┌─────────┐
 │         │                │         │              │         │
 │  DEV-A  ├────────────────►  DEV-B  ├──────────────►  DEV-C  │
 │         │                │         │              │         │
 └─────────┘                └─────────┘              └─────────┘
</pre>
However, Dev-C turns into a GM as well. We figured out that the announce messages sent from Dev-B inside contain wrong information. It includes information from its own configuration instead of forwarding the information about the GM. In that case Dec-C assumes having the better clock and therefore turns into a master as well. This results in a network topology having two GM.

The expected behavior is that Dev-B forwards the clock information from Dev-A via  announce messages on its other ports. 

Whil eanalyzing the problem we made some changes to *gptp_messages.c* as shown below what seems to solve the problem.

```diff
------------------------- gptp_messages.c -------------------------
index 06b9cb0..f6cf1c9 100644
@@ -502,6 +502,7 @@ struct net_pkt *gptp_prepare_announce(int port)
 	struct net_if *iface;
 	struct net_pkt *pkt;
 	struct gptp_hdr *hdr;
+	struct gptp_priority_vector *gm_prio;
 
 	NET_ASSERT((port >= GPTP_PORT_START) && (port <= GPTP_PORT_END));
 	global_ds = GPTP_GLOBAL_DS();
@@ -551,25 +552,12 @@ struct net_pkt *gptp_prepare_announce(int port)
 	ann->cur_utc_offset = htons(global_ds->current_utc_offset);
 	ann->time_source = global_ds->time_source;
 
+	gm_prio = &global_ds->gm_priority;
 	switch (GPTP_PORT_BMCA_DATA(port)->info_is) {
 	case GPTP_INFO_IS_MINE:
-		ann->root_system_id.grand_master_prio1 = default_ds->priority1;
-		ann->root_system_id.grand_master_prio2 = default_ds->priority2;
-
-		ann->root_system_id.clk_quality.clock_accuracy =
-			default_ds->clk_quality.clock_accuracy;
-		ann->root_system_id.clk_quality.clock_class = default_ds->clk_quality.clock_class;
-		ann->root_system_id.clk_quality.offset_scaled_log_var =
-			htons(default_ds->clk_quality.offset_scaled_log_var);
-
-		memcpy(&ann->root_system_id.grand_master_id,
-		       default_ds->clk_id,
-		       GPTP_CLOCK_ID_LEN);
-		break;
 	case GPTP_INFO_IS_RECEIVED:
 		memcpy(&ann->root_system_id,
-		       &GPTP_PORT_BMCA_DATA(port)->
-				master_priority.root_system_id,
+		       &gm_prio->root_system_id,
 		       sizeof(struct gptp_root_system_identity));
 		break;
 	default:
```

### PR DESCRIPTION
**Pull request by mschappa**
Fixes #66646

### CODE REVIEW

**Comment by jukkar** [comment]
Please add signed-off-by to the commit message. Also add `Fixes #66646` to the commit message so that the original issue gets closed automatically.
You had nice information about the problem in the issue, at least some of that text could be placed to the commit message to give more information why this change is needed.

**Comment by mschappa** [comment]
> Please add signed-off-by to the commit message. Also add `Fixes: #66646` to the commit message so that the original issue gets closed automatically. You had nice information about the problem in the issue, at least some of that text could be placed to the commit message to give more information why this change is needed.

done

**Comment by mariopaja** [comment]
> cc @talih0 @mariopaja please take a look

@jukkar I am out of office at the moment. I will not be able to test it out on the setup with the managed gPTP switch before the new year. However, I will test it once on a peer-to-peer connection

File: subsys/net/l2/ethernet/gptp/gptp_messages.c [hunk 1]
```diff
@@ -538,25 +539,12 @@ struct net_pkt *gptp_prepare_announce(int port)
...
-			htons(default_ds->clk_quality.offset_scaled_log_var);
-
-		memcpy(&ann->root_system_id.grand_master_id,
-		       default_ds->clk_id,
-		       GPTP_CLOCK_ID_LEN);
-		break;
 	case GPTP_INFO_IS_RECEIVED:
 		memcpy(&ann->root_system_id,
-		       &GPTP_PORT_BMCA_DATA(port)->
-				master_priority.root_system_id,
```
Comment (talih0) [inline subsys/net/l2/ethernet/gptp/gptp_messages.c:559]: @mschappa, thanks, the change looks fine to me.

I'm trying to figure out why DevB is not forwarding the announce message properly at the moment. Shouldn't master_priority.root_system_id already be same as gm_prio->root_system_id at this point? 
Or is it getting stuck in the GPTP_INFO_IS_MINE case?

Would you mind uploading a wireshark capture before and after the change, please?

Comment (mschappa) [reply subsys/net/l2/ethernet/gptp/gptp_messages.c:559]: Hi @talih0, thanks for taking a look. At the moment I do not have access to the board but I can do a capture during next week probably. While debugging, for me it looked like it is always considering GPTP_INFO_IS_MINE case before sending announce messages and therefore, the information always contained the configured values instead the once received from the GM.

Comment (jukkar) [reply subsys/net/l2/ethernet/gptp/gptp_messages.c:559]: @mschappa any progress on this?

Comment (mschappa) [reply subsys/net/l2/ethernet/gptp/gptp_messages.c:559]: @jukkar @talih0 sorry for the delay. I was out of office until today.

Please find attached two wireshark logs _announce_before_fix.pcap_ and _announce_after_fix.pcap_ within [announce_pcaps.zip](https://github.com/zephyrproject-rtos/zephyr/files/13888173/announce_pcaps.zip). Both have been captured at DEV-C according to the issue description #66646:
 - Dev-C is identified as LCFCHeFe_7c:a9:ac
 - Dev-B is identified as RenesasE_00:00:01

When looking a the announce messages before the fix, we can see that DEV-B forwards its own priority1 as 200. After the fix it forwards the GM prioity1 as 100.

grafik
grafik

Comment (talih0) [reply subsys/net/l2/ethernet/gptp/gptp_messages.c:559]: thanks, @mschappa, I'll look over tomorrow.

**Review by talih0 — APPROVED** [review]
Thanks, this looks good to me.

**Review by jukkar — APPROVED** [review]

**Pull request merged**